### ISSUE TITLE
Update netcdf configuration.

### ISSUE THREAD

**Issue by roystgnr** [libMesh/libmesh#721]
Aside from the "our netcdf is out of date" issue in #363 and the "our netcdf may not build with some compilers" issue in #602, @dmcdougall now seems to have run into a "our netcdf may conflict with VTK's netcdf" issue.  The first two cases might be fixed by upgrading our contrib version to 4.3.3.1, but for the last case I don't think there's a robust workaround other than "build netcdf separately, tell VTK to use that build, and tell libMesh to use that build too".

**Comment by jwpeterson** [comment]
Ugh.  Out of curiosity, what cmake flag do you set to make VTK require its own NetCDF, and what additional features does having NetCDF in VTK enable?

**Comment by roystgnr** [comment]
I had set nothing, and I have no idea.

Looks like there's a vtkNetCDFReader class?

**Comment by pbauman** [comment]
Yeah, at some point in the newer versions of VTK, it, by default, will fetch and internally build NetCDF for its own purposes.

**Comment by roystgnr** [comment]
Any idea how to disable it?  I see a NETCDF_ENABLE_NETCDF4 cmake bool, but I can't tell whether turning that off would leave you with nothing or with NetCDF 3.

**Comment by jwpeterson** [comment]
Newer than VTK 6.3?  I have not seen any incompatibility issues up to this version...

**Comment by roystgnr** [comment]
VTK 5.10.0

But here's what's crazy: whether by wit or luck or magic, libtool works around the issue.  When you say "libtool --etc g++ my_link_instructions", those link instructions get rejiggered and reordered, and libtool executes (and echos to stdout) "g++ altered_link_instructions".  Run g++ my_link_instructions manually and you get an executable that dies as soon as you try to write an ExodusII file.  Run with libtool or run g++ altered_link_instructions, and everything works.

**Comment by jwpeterson** [comment]
> VTK 5.10.0

Yikes.  I'm glad to know that version still works, I guess.



> But here's what's crazy: whether by wit or luck or magic, libtool works around the issue.

Interesting, but I vote for luck as the reason.

**Comment by roystgnr** [comment]
I've got to agree with "luck".

**Comment by roystgnr** [comment]
Okay, we can add "doesn't work with the newest HDF5" to our list of libMesh contrib NetCDF woes.

http://www.unidata.ucar.edu/mailing_lists/archives/netcdfgroup/2014/msg00188.html
https://www.hdfgroup.org/HDF5/doc/RM/H5P/H5Pset_fapl_mpiposix.htm

**Comment by jwpeterson** [comment]
On Fri, May 27, 2016 at 1:58 PM, roystgnr <notifications@github.com> wrote:

> Okay, we can add "doesn't work with the newest HDF5" to our list of
> libMesh contrib NetCDF woes.
>
>
> http://www.unidata.ucar.edu/mailing_lists/archives/netcdfgroup/2014/msg00188.html
> https://www.hdfgroup.org/HDF5/doc/RM/H5P/H5Pset_fapl_mpiposix.htm
>

Are you sure about this?  I tested it a while back and couldn't get it to
fail.  There is a configure test in NetCDF that seems like it would take
care of this.  Note that I think we have two NETCDFs in libmesh...make sure
you are using the newer one.

-- 
John

**Comment by roystgnr** [comment]
It hit me on lonestar 5, using the TACC phdf5/1.8.16 module:

```../liblib/.libs/libnetcdf.so: undefined reference to `H5Pset_fapl_mpiposix'```

**Comment by roystgnr** [comment]
And yes, I was configured to use netcdf v4.

**Comment by jwpeterson** [comment]
Here's the relevant excerpt from contrib/netcdf/v4/configure.ac:

```
   AC_CHECK_FUNCS([H5Pget_fapl_mpiposix H5Pget_fapl_mpio H5Pset_deflate H5Z_SZIP])

   # The user may have parallel HDF5 based on MPI POSIX.
   if test "x$ac_cv_func_H5Pget_fapl_mpiposix" = xyes; then
      AC_DEFINE([USE_PARALLEL_POSIX], [1], [if true, compile in parallel netCDF-4 based on MPI/POSIX])
   fi

   # The user may have parallel HDF5 based on MPI mumble mumble.
   if test "x$ac_cv_func_H5Pget_fapl_mpio" = xyes; then
      AC_DEFINE([USE_PARALLEL_MPIO], [1], [if true, compile in parallel netCDF-4 based on MPI/IO])
   fi

   # If parallel is in use, enable it in the C code. Also add some stuff to netcdf.h.
   enable_parallel=no
   if test "x$ac_cv_func_H5Pget_fapl_mpio" = xyes -o "x$ac_cv_func_H5Pget_fapl_mpiposix" = xyes; then
      enable_parallel=yes
      AC_DEFINE([USE_PARALLEL], [1], [if true, parallel netCDF-4 is in use])
   fi
```

So if neither function is available, USE_PARALLEL should not be defined.  I also looked briefly through the source and did not find a place where it was used outside of USE_PARALLEL guards, but even if it was, it seems like it would be a compiling error and not a linking error, if the function is not available on the system...

**Comment by jwpeterson** [comment]
NetCDF version was updated in #1326. This ticket seems to also be about VTK's use of NetCDF and the interaction between libmesh, VTK, and NetCDF, so I'll leave it open for the time being...

**Comment by mboisson** [comment]
Why is there no option to use a pre-existing install of NetCDF ? I also have a conflict when building libMesh because it builds NetCDF in a way that conflicts with ours.
